lib/membrane_rtc_engine/endpoints/webrtc/simulcast_tee.ex

```diff
@@ -147,8 +149,48 @@ defmodule Membrane.RTC.Engine.Endpoint.WebRTC.SimulcastTee do
         _other -> []
       end)
 
-    start_timer = [start_timer: {:check_encoding_statuses, Time.seconds(1)}]
-    {{:ok, start_timer ++ caps}, state}
+    {{:ok, caps}, state}
+  end
+
+  @impl true
+  def handle_event(%TrackVariantPaused{}, encoding, ctx, state) do
+    new_state = update_in(state, [:inactive_encodings], &MapSet.put(&1, encoding))
+
+    new_state =
+      Enum.reduce(state.forwarders, new_state, fn {endpoint_id, forwarder}, new_state ->
+        put_in(
+          new_state,
+          [:forwarders, endpoint_id],
+          Forwarder.encoding_inactive(forwarder, encoding)
+        )
+      end)
+
+    actions = generate_keyframe_requests(new_state, state, ctx)
+
+    {actions, new_state}
+  end
+
+  @impl true
+  def handle_event(%TrackVariantResumed{}, encoding, ctx, state) do
+    new_state = update_in(state, [:inactive_encodings], &MapSet.delete(&1, encoding))
+
+    new_state =
+      Enum.reduce(state.forwarders, new_state, fn {endpoint_id, forwarder}, new_state ->
+        put_in(
```

Comment:
These functions looks very similar, could we merge it?

Reply:
I'd add one more to the list - lines 76-82 that have a nicer formatting :wink:

Reply:
The only thing that cames to my mind is something like this

```elixir
  @impl true
  def handle_event(Pad.ref(:input, {_track_id, encoding}), %m{} = event, ctx, state)
      when m in [TrackVariantResumed, TrackVariantPaused] do
    {forwarder_update_fn, inactive_encodings_update_fn} =
      case event do
        %TrackVariantPaused{} -> {&Forwarder.encoding_inactive/2, &MapSet.delete/2}
        %TrackVariantResumed{} -> {&Forwarder.encoding_active/2, &MapSet.put/2}
      end

    new_state =
      update_in(state, [:inactive_encodings], &inactive_encodings_update_fn.(&1, encoding))

    new_state =
      Enum.reduce(state.forwarders, new_state, fn {endpoint_id, forwarder}, new_state ->
        forwarder = forwarder_update_fn.(forwarder, encoding)
        put_in(new_state, [:forwarders, endpoint_id], forwarder)
      end)

    actions = generate_keyframe_requests(new_state, state, ctx)

    {actions, new_state}
  end
```

But I don't think it's better. Those two events are similar but we do opposite things when handling them

Reply:
Hmm, IMO this doesn't look bad.

Reply:
OK, I guess with @Rados13 we meant a bit different things. I'm not sure if the merge of actual handle_event clauses is needed, but the common part is the state updates that I'd extract as state-modifying functions, one for updating forwarders and one for inactive encodings. Then, you could reuse them in L76-82. Something like:
* `set_encoding_active(state, bool)`
* `set_forwarder_active(state, bool)`

Reply:
Sounds good :+1:  Let me introduce this idea in the next PR as I am going to modify `handle_event` callbacks there.

Reply:
> OK, I guess with @Rados13 we meant a bit different things. I'm not sure if the merge of actual handle_event clauses is needed, but the common part is the state updates that I'd extract as state-modifying functions, one for updating forwarders and one for inactive encodings. Then, you could reuse them in L76-82. Something like:
> 
> * `set_encoding_active(state, bool)`
> * `set_forwarder_active(state, bool)`

Ok, this way this looks even better.